File: Easy/ReverseLinkedLists.cpp
/*
  206. Reverse Linked Lists

  Given the head of a singly linked list, reverse the list, and return the
  reversed list.

  Example 1:
  Input: head = [1,2,3,4,5]
  Output: [5,4,3,2,1]
 */

struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
  // Reassigns the ->next property of each NodeList as it iterates along. Since
  // it loops through only once, this algorithm has a linear time complexity.
  // Time Complexity O(n).
  ListNode *reverseList(ListNode *head) {
    // Corner case if input list is null or only has one element
    if (head == nullptr || head->next == nullptr) {
      return head;
    }
    // Initialise new ListNode to track reversed list, as well as helper
    // pointers to reassign elements
    ListNode *reversed = nullptr;
    ListNode *current = head;
    ListNode *next = nullptr;
    // Reverse the .next value of each linked list
    while (current != nullptr) {
      next = current->next;
      current->next = reversed;
      reversed = current;
      current = next;
    }
    return reversed;
  }
};

File: Easy/ReverseLinkedLists.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution(object):
    def reverseList(self, head):
        """
        Given the head of a singly linked list, reverse the list,
        and return the reversed list. Both algorithms have time
        complexity of O(n), but the space complexity is vastly
        different.

        :type head: ListNode
        :rtype: ListNode
        """
        if not head:
            return

        previous = None

        while head is not None:
            newList = ListNode(head.val, previous)
            previous = newList
            head = head.next

        return previous

    def reverseList(self, head):
        """
        Given the head of a singly linked list, reverse the list,
        and return the reversed list.

        :type head: ListNode
        :rtype: ListNode
        """
        if not head:
            return

        previous = None
        current = head
        while head is not None:
            head = head.next
            current.next = previous
            previous = current
            current = head
        return previous
